### PR TITLE
Compatibility with MacOS and C++17

### PR DESCRIPTION
**Pull request by mholtrop**
BEGINRELEASENOTES
- Make compatible with MacOS Monterey and Ventura
- Make compatible with C++17 - requires newer compiler. Tested with gcc 9.3.0 on Centos 7.
ENDRELEASENOTES

### CODE REVIEW

**Review by sarahgaiser — COMMENTED** [review]
Do we have to remove all the throw (...) statements to make this compile on macOS?

**Comment by mholtrop** [comment]
Yes. The throw statements in the declaration of a method in C++ was never useful. It was made deprecated with C++11, 12 years ago, and removed from c++17.

File: src/cpp/include/SIO/SIOReader.h [hunk 1]
```diff
@@ -51,56 +51,55 @@ class SIOEventHandler ;
      * user.
      * @throws IOException
      */
-    virtual void open(const std::vector<std::string>& filenames) 
-      throw (IO::IOException, std::exception) ;
+    virtual void open(const std::vector<std::string>& filenames);
 
 
     /** Opens a file for reading (read-only).
      * @throws IOException
```
Comment (cbravo135) [inline src/cpp/include/SIO/SIOReader.h:58]: We should update the comments too, this no longer throws and IOException. There are a few places where the comments can be updated.

Comment (mholtrop) [reply src/cpp/include/SIO/SIOReader.h:58]: Unless I am really wrong about C++ method declarations, the "throw()" in the method declaration, or the absence of it, does not actually influence whether the code throws an exception or not. It was "decorative", in telling the compiler that this method might throw an exception, but never really taken into account by compilers.
If you are sure the code does not throw an IOException anywhere any longer, then yes, we can remove that comment.

File: src/cpp/CMakeLists.txt [hunk 2]
```diff
@@ -453,19 +453,17 @@ MACRO( ADD_LCIO_EXAMPLE file )
...
-ADD_LCIO_EXAMPLE( stdhepjob ) 
-ADD_LCIO_EXAMPLE( stdhepjob_new )
+ADD_LCIO_EXAMPLE( copyfix )
 ADD_LCIO_EXAMPLE( addRandomAccess )
 ADD_LCIO_EXAMPLE( lcio_event_counter ) 
 ADD_LCIO_EXAMPLE( lcio_check_col_elements )
 ADD_LCIO_EXAMPLE( lcio_split_file ) 
 ADD_LCIO_EXAMPLE( lcio_merge_files ) 
-ADD_LCIO_EXAMPLE( readmcparticles ) 
+#ADD_LCIO_EXAMPLE( readmcparticles )
```
[the remark this answers is not in the document]
Comment (mholtrop) [reply src/cpp/CMakeLists.txt:467]: If you would like to maintain the code for the examples, great, but I don't think that the "readmcparticles" example code should be stopping us from making the modifications in this pull. Just my opinion.

File: src/cpp/include/SIO/SIOReader.h (hunk 1 above)
Comment (cbravo135) [reply src/cpp/include/SIO/SIOReader.h:58]: I am pretty sure this was a way to make an "interface", with the idea that these exceptions were thrown if these methods were not overloaded by the inheriting class but were called. This particular implementation of the class no longer throws the exception anymore and therefore should not claim to throw the exception and comments should be added where it is actually throw if ever thrown elsewhere.

File: src/cpp/CMakeLists.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (cbravo135) [reply src/cpp/CMakeLists.txt:467]: IMO we should discuss and coordinate any changes to the list of targets in the build if ever changed. I haven't blocked this PR, but will only give my approval in one of two ways on this, either we discuss why we remove the targets or we don't make a change to the list of targets. I am happy with going either direction from here and this should be decided by the source of the PR.

**Comment by cbravo135** [comment]
@mholtrop I did pull these commits locally and tested building and running various things and am pretty convinced this is safe overall. I do want to make discuss a bit (which can be done here) and clean up the comments, but besides that I am happy with this PR. During the last weekly meeting you complained about the amount of coordination happening in this repo, and I am simply putting effort into improving that situation.

**Comment by mholtrop** [comment]
@cbravo135 Hello Cam, Sarah made the changes in the comments and restored the examples in the CMakeLists file. I hope this is now in shape to be merged. Thanks.

**Review by sarahgaiser — APPROVED** [review]

**Pull request merged**